### PR TITLE
Propagate_labels: implement and verify Kani harnesses (6.4.6)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Refactors and extends Kani verification for propagate_labels: adds a modular harness layout, exposes a crate-visible report, and introduces tests and an ExecPlan for roadmap item 6.4.6.
- Splits the previous monolithic 6.4.5/6.4.6 harness into a crate-scoped structure under
  common/src/decomposition_advice/community_kani/ with:
  - adjacency.rs, propagate_labels.rs, shared.rs, mod.rs
- Replaces the old community_kani.rs harness and preserves the 6.4.5 pattern while adding 6.4.6 focused tests and proofs.
- Adds unit tests and a test-support seam for label propagation, plus BDD-style behaviour tests scaffolding.
- Adds an ExecPlan documenting 6.4.6: docs/execplans/6-4-6-kani-verification-propagate-labels-preserves-indices.md
- Updates design/docs to reflect decisions and gating for 6.4.6 (brain-trust-lints-design.md).
- Adds test-support surface for label propagation and behaviour tests:
  - common/src/test_support/decomposition_label_propagation.rs
  - common/tests/features/decomposition_label_propagation.feature
  - common/tests/decomposition_label_propagation_behaviour.rs
- Fixes a bug in scripts/install-kani.sh cleanup so warm caches don’t cause CI to fail.
- No public API changes beyond internal test/support seams; runtime behaviour preserved.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---
## Changes
- Added new Kani harness modules:
  - common/src/decomposition_advice/community_kani/adjacency.rs
  - common/src/decomposition_advice/community_kani/propagate_labels.rs
  - common/src/decomposition_advice/community_kani/shared.rs
  - common/src/decomposition_advice/community_kani/mod.rs
- Removed old harness aggregator: common/src/decomposition_advice/community_kani.rs
- Updated propagation runtime:
  - common/src/decomposition_advice/community.rs now includes LabelPropagationReport and propagate_labels_report, and propagate_labels now returns the final labels via into_labels()
- Updated crate exports:
  - common/src/decomposition_advice/mod.rs to re-export minimal_feature_vector alongside existing exports
- Tests and test-support:
  - common/src/decomposition_advice/tests/propagation.rs (unit tests for propagation)
  - common/src/test_support/decomposition_label_propagation.rs (runtime-visible label-propagation seam for tests)
  - common/tests/features/decomposition_label_propagation.feature (BDD scenarios)
  - common/tests/decomposition_label_propagation_behaviour.rs (behaviour tests harness)
- ExecPlan and docs:
  - docs/execplans/6-4-6-kani-verification-propagate-labels-preserves-indices.md (new)
  - docs/brain-trust-lints-design.md updated with 6.4.6 decisions and approach
- Test support and surfaces:
  - common/src/test_support/decomposition.rs extended to expose LabelPropagationReport from test support
- Behavioural and feature test wiring:
  - common/src/test_support/decomposition_label_propagation.rs provides a runtime-visible LabelPropagationReport surface for tests
- CI and tooling:
  - scripts/install-kani.sh: cleanup path fixed to not fail on warm caches
- Documentation task links preserved in the PR body

---
## Rationale
- Keeps 6.4.5 sidecar approach and gradually introduces 6.4.6 harnesses with a small, split-proof surface to avoid repo bloat and state explosion in Kani.
- Introduces a crate-visible propagation report to enable unit tests, Kani, and behaviour tests to observe a stable runtime contract without widening production APIs.
- Provides a solid test-support seam to exercise label propagation through declarative inputs while containing unhappy-path coverage within test-support rather than production code.

---
## Validation plan (per ExecPlan)
- Unit tests target: length of labels, labels in range, isolated nodes, zero-iteration behavior, lexical tie-breaking, and bounded non-convergence behavior.
- Kani harnesses (bounded model checking) exercise a small, fixed-node graph with symbolic edges/weights, ensuring termination within a caller-provided max_iterations and acceptance of valid inputs.
- Behaviour tests scaffolded via rstest-bdd for declarative graph inputs through the test-support surface; verify observable outcomes without widening runtime APIs.
- Documentation and roadmap updates accompany changes; CI gates continue to exercise fmt, lint, tests, and Kani where possible.

---
## Next steps
- After this PR lands, implement the A–F stages in the ExecPlan, including any necessary file splits, refinements to test-support helpers, and updates to documentation.
- Iterate on Kani gate stability and keep design docs in sync with implementation progress.

---
## Review guidance
- Ensure the ExecPlan clearly scopes to 6.4.6 and references the existing 6.4.5 sidecar.
- Validate the validation plan (unit tests, BDD tests, and Kani harnesses) is coherent and achievable within repository constraints.
- Check for missing prerequisites or gating criteria that could block progress.

---
## Notes
- This PR introduces executable test support and Kani harness structure for propagate_labels without changing public APIs or runtime behaviour.

---
## Task
ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/35067eb9-9001-41a3-b6e8-d273c148d83d

📎 **Task**: https://www.devboxer.com/task/96058d73-2f64-4ccb-a5a1-e1b1519a2f59

## Summary by Sourcery

Introduce a crate-visible label propagation report surface, expand verification and behavioural coverage for propagate_labels, and fix a Kani installer cleanup bug while keeping public APIs and runtime behaviour unchanged.

New Features:
- Expose a crate-visible LabelPropagationReport and propagate_labels_report helper to surface label propagation results for internal consumers and tests.
- Add a dedicated test-support label_propagation_report API and report type to drive declarative label propagation behaviour tests.
- Introduce new Kani harnesses for propagate_labels using a bounded symbolic graph model.

Bug Fixes:
- Fix scripts/install-kani.sh cleanup so warm cached installs no longer cause the Kani sidecar to exit with a failure status.

Enhancements:
- Refine label tie-breaking in propagate_labels to use method_name() then label index, preserving determinism while simplifying comparison for verification.
- Refactor Kani verification for decomposition community helpers into a modular community_kani module with separate adjacency and propagate_labels harness files.
- Add unit tests that pin the observable label propagation contract, including bounds, isolated nodes, zero-iteration behaviour, lexical tie-breaking, and bounded non-convergence.
- Extend decomposition_advice test support to expose label propagation reporting and reuse adjacency validation helpers.
- Expose a minimal_feature_vector constructor for building lightweight MethodFeatureVector values in tests and proofs.

Documentation:
- Add an ExecPlan documenting roadmap item 6.4.6 for Kani verification of propagate_labels and its validation strategy.
- Update brain-trust-lints-design.md with the new label propagation report seam, tie-breaking behaviour, Kani harness layout, and known verification limitations.

Tests:
- Add BDD-style behaviour tests and feature definitions for label propagation scenarios, including disconnected components, isolated graphs, zero-iteration inputs, lexical tie-breaking, and invalid input rejection.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- In `best_neighbour_label` the `scores` vector is sized to `labels.len()` on every call, which can be significantly larger than the number of neighbours; consider using a more tightly scoped structure (e.g., a small map keyed by neighbour labels or a scratch buffer reused across calls) to avoid unnecessary allocations and potential cache pressure on large graphs.
- There are now two separate `LabelPropagationReport` types (in `community.rs` and in `test_support/decomposition_label_propagation.rs`) that mirror each other; to reduce the risk of these diverging over time, consider reusing the core type directly or introducing a thin wrapper/adapter instead of duplicating fields and accessors.

## Individual Comments

### Comment 1
<location path="common/src/decomposition_advice/community.rs" line_range="107" />
<code_context>
     communities
 }

+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct LabelPropagationReport {
+    labels: Vec<usize>,
</code_context>
<issue_to_address>
**issue (complexity):** Consider simplifying the label propagation reporting and comparison helpers to avoid extra state and indirection while keeping the same observable behaviour and testability.

You can keep the new behaviour while trimming some indirection and state.

**1. Simplify `LabelPropagationReport` and the wrapper**

`has_active_nodes` is derivable by the caller (they already have `adjacency`) and adds state that can go stale. Also, `into_labels` + thin wrapper adds surface without much gain.

You can keep the report for tests while making it smaller and more direct:

```rust
#[derive(Clone, Debug, Eq, PartialEq)]
pub(crate) struct LabelPropagationReport {
    pub(crate) labels: Vec<usize>,
    pub(crate) iteration_count: usize,
}

pub(crate) fn propagate_labels(
    vectors: &[MethodFeatureVector],
    adjacency: &[Vec<(usize, u64)>],
    max_iterations: usize,
) -> Vec<usize> {
    propagate_labels_report(vectors, adjacency, max_iterations).labels
}

pub(crate) fn propagate_labels_report(
    vectors: &[MethodFeatureVector],
    adjacency: &[Vec<(usize, u64)>],
    max_iterations: usize,
) -> LabelPropagationReport {
    let mut labels: Vec<usize> = (0..vectors.len()).collect();
    let active_nodes: Vec<_> = adjacency
        .iter()
        .enumerate()
        .filter_map(|(node, neighbours)| (!neighbours.is_empty()).then_some(node))
        .collect();

    if active_nodes.is_empty() {
        return LabelPropagationReport {
            labels,
            iteration_count: 0,
        };
    }

    let mut iteration_count = 0;
    for _ in 0..max_iterations {
        iteration_count += 1;
        let mut changed = false;

        for &node in &active_nodes {
            let Some(best_label) =
                best_neighbour_label(node, &labels, adjacency, vectors)
            else {
                continue;
            };

            if best_label != labels[node] {
                labels[node] = best_label;
                changed = true;
            }
        }

        if !changed {
            break;
        }
    }

    LabelPropagationReport {
        labels,
        iteration_count,
    }
}
```

Callers that need `has_active_nodes` can compute:

```rust
let has_active_nodes = adjacency.iter().any(|n| !n.is_empty());
```

This keeps all functionality (labels + iteration count, early return when no active nodes) while removing extra state and `into_labels()`.

**2. Collapse the comparison helpers**

`should_replace_best` → `is_better_label` → `breaks_label_tie` spreads simple logic over three functions. You can inline the tie-breaking into one function with a short comment, making the choice rule easier to read in one place:

```rust
fn should_replace_best(
    current_best: Option<(usize, u64)>,
    candidate_label: usize,
    candidate_score: u64,
    vectors: &[MethodFeatureVector],
) -> bool {
    match current_best {
        None => true,
        Some((best_label, best_score)) => {
            // Prefer higher score; on tie, lexicographically smaller name;
            // if names equal, smaller label index for determinism.
            if candidate_score != best_score {
                candidate_score > best_score
            } else {
                let candidate_name = vectors[candidate_label].method_name();
                let best_name = vectors[best_label].method_name();

                candidate_name < best_name
                    || (candidate_name == best_name && candidate_label < best_label)
            }
        }
    }
}
```

Then `best_neighbour_label` stays as-is, just calling `should_replace_best` directly.

These changes keep deterministic tie-breaking and the richer reporting for tests, while reducing the number of layers and state you have to keep in your head when reading the label propagation logic.
</issue_to_address>

### Comment 2
<location path="docs/execplans/6-4-6-kani-verification-propagate-labels-preserves-indices.md" line_range="243" />
<code_context>
+- Decision: stop further Kani-model churn after three propagation-model
+  iterations. Rationale: the ExecPlan tolerance says to escalate once the proof
+  still fails after repeated bounded-model simplification attempts; the current
+  blocker is in CBMC container machinery, not in the propagation property
+  logic. Date/Author: 2026-04-14 / Codex.
+
</code_context>
<issue_to_address>
**issue (review_instructions):** The acronym “CBMC” is used here without being expanded on first use in this document, which violates the requirement to define uncommon acronyms.

Consider expanding CBMC the first time it appears in this ExecPlan, for example:

"blocker is in C Bounded Model Checker (CBMC) container machinery, not in the propagation property"

Subsequent uses can then keep just "CBMC".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: common/src/decomposition_advice/community.rs [hunk 1]
```diff
@@ -167,16 +213,16 @@ fn best_neighbour_label(
         return None;
     }
 
-    let mut scores: BTreeMap<usize, u64> = BTreeMap::new();
+    let mut scores = vec![0; labels.len()];
```
[the remark this answers is not in the document]
Comment (leynos) [reply common/src/decomposition_advice/community.rs]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

File: common/src/decomposition_advice/community.rs [hunk 2]
```diff
@@ -122,19 +139,56 @@ pub(crate) fn build_adjacency(
...
+/// assumes validated adjacency input and may panic if a neighbour index falls
+/// outside that range.
+///
+/// If `adjacency` contains no active nodes, or if `max_iterations` is `0`, it
+/// returns the initial self labelling with an `iteration_count` of `0`.
+pub(crate) fn propagate_labels_report(
+    vectors: &[MethodFeatureVector],
+    adjacency: &[Vec<(usize, u64)>],
+    max_iterations: usize,
+) -> LabelPropagationReport {
```
[the remark this answers is not in the document]
Comment (leynos) [reply common/src/decomposition_advice/community.rs:170]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 errors, 2 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | BDD scenario tests are documented placeholders with empty bodies; unit tests lack convergence detection, score accumulation, and edge-case coverage. | Implement five BDD scenario functions to invoke feature steps; add convergence tests, boundary cases, and expanded tie-breaking coverage. |
| Module-Level Documentation | ❌ Error | Public function `minimal_feature_vector` in `common/src/decomposition_advice/vector.rs` lacks required docstring documentation. | Add docstring to `minimal_feature_vector` explaining it constructs a minimal `MethodFeatureVector` with supplied method name and empty feature maps, with usage example. |
| Developer Documentation | ⚠️ Warning | Pull request adds public test-support APIs (label\_propagation\_report and LabelPropagationReport) but fails to document them in developers-guide.md despite established adjacency API documentation pattern. | Add section to docs/developers-guide.md after line 323 documenting label\_propagation\_report function, LabelPropagationReport struct, observable methods, validation semantics, and ExecPlan cross-reference. |
| Testing (Unit) | ⚠️ Warning | Unit tests are substantive and cover meaningful behaviour, edge cases, and error paths. Behavioural test scenario functions are empty placeholders that do not exercise the functional boundary. | Implement the five scenario functions in common/tests/decomposition\_label\_propagation\_behaviour.rs to invoke step handlers and execute feature scenarios with assertions. |
| Testing (Property / Proof) | ❓ Inconclusive | Cannot execute shell commands to examine Kani harness files and test implementations. | Provide the file contents directly or use web search to find publicly available Kani verification documentation. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Unit tests assert concrete label outcomes but use uniform weights throughout, creating a gap where count-based implementations would pass identically despite violating weight-based scoring requirements. | Introduce heterogeneous-weight unit tests and extend Kani harnesses to assert concrete label outcomes for specific weight distributions, ensuring weight-based scoring is verified as critical. |
| Domain Architecture | ⚠️ Warning | minimal\_feature\_vector() exported as pub(crate) lacks #\[cfg(test)\] or #\[cfg(any(test, kani))\] guard despite exclusive use in testing, violating domain architecture and blurring test-production boundaries. | Apply #\[cfg(any(test, kani))\] to minimal\_feature\_vector() in common/src/decomposition\_advice/vector.rs and remove the re-export from mod.rs. |
| Observability | ⚠️ Warning | Iteration count computed by propagate\_labels\_report is immediately discarded in production code; diagnostic convergence data has no visibility to maintainers. | Log iteration count, node count, and convergence status at the infer\_communities call site using info or debug level logging. |

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational.

Check name | Status | Explanation | Resolution
-- | -- | -- | --
Testing (Overall) | ❌ Error | The propagate_labels wrapper function in community.rs is dead code—no test, Kani harness, or production code calls it, contradicting its #[expect(dead_code)] claim of being a test convenience. | Delete the dead code function or add actual test/Kani callsites with #[cfg(any(test, kani))] gating and update the dead-code expectation to match reality. Address the reviewer's flagged dead code concern before approval.
Module-Level Documentation | ❌ Error | Function propagate_labels() carries documentation  falsely claiming it serves as a convenience seam for tests and  harnesses, but zero callers exist; all tests use  propagate_labels_report() instead. | Remove propagate_labels() entirely or restore genuine  callers and rewrite documentation to acknowledge its dead-code status  without false claims about actual usage.
Unit Architecture | ❌ Error | Function propagate_labels is decorated with a false #[expect(dead_code)]  claim of serving as a convenience seam, yet comprehensive repository  search reveals zero call sites anywhere—tests, Kani harnesses, and  runtime code all call propagate_labels_report directly instead. | Remove propagate_labels entirely or restore at least one genuine production/test call site and update the #[expect] justification to match reality.
Developer Documentation | ⚠️ Warning | Roadmap item 6.4.6 remains incomplete despite code  submission; propagate_labels wrapper is dead code with zero callers  despite false documentation claim; Makefile PATH changes lack  architectural documentation. | Check off roadmap item 6.4.6 after removing  propagate_labels or establishing legitimate non-test callers. Document  Makefile PATH tooling changes in developers-guide.md. Update  brain-trust-lints-design.md with final disposition of propagate_labels.
Testing (Unit) | ⚠️ Warning | Unit tests verify local behaviour effectively, but PR fails integration testing: dead propagate_labels carries false seam justification, and BDD tests exercise only test-support seams rather than production code paths through detect_communities. | Remove dead propagate_labels function and add integration tests exercising detect_communities with representative inputs to verify refactored propagation works end-to-end through the production pipeline.
Domain Architecture | ⚠️ Warning | The PR introduces dead code (propagate_labels) with false justification and blurs domain boundaries by duplicating label propagation logic in test-support. | Remove unreachable propagate_labels().  Consolidate test-support wrappers to avoid replicating domain  invariants; ensure domain layer owns all label propagation authority.
Observability | ⚠️ Warning | propagate_labels_report lacks logging at critical  decision points: empty active nodes, early convergence, and iteration  limit reached. Only caller-level DEBUG logging exists, violating PR  objectives for diagnostic logging within the core function. | Add structured logging directly within  propagate_labels_report at all decision boundaries with node count,  active node count, iteration count, and convergence state. Replace  assertion panic with logged error.
